S3 cache doc updates v2 (#1018)
harness/developer-hub: docs/continuous-integration/use-ci/caching-ci-data/saving-cache.md

helpdocs_is_published: true
---

```mdx-code-block
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
```

> Caching enables sharing data across stages. Caching also speeds up builds by reusing the expensive fetch operation from previous jobs.

Reviewer: ```suggestion
Caching enables sharing data across stages. Caching also speeds up builds by reusing data from expensive fetch operations in previous jobs.
```

Caching has two primary benefits:

> 1. Run pipelines faster by reusing the expensive fetch operation data from previous builds
> 2. Share data across stages

Reviewer: ```suggestion
* Run pipelines faster by reusing the expensive fetch operation data from previous builds
* Share data across stages
```

> You can cache data to an AWS S3 bucket in one Stage using the **Save Cache to S3** step, and restore it in the same Stage, or a following Stage, using **Restore Cache From S3** step. 

Reviewer: ```suggestion
You can cache data to an AWS S3 bucket in one stage using the **Save Cache to S3** step, and restore it in the same stage or a following stage using the **Restore Cache From S3** step. 
```

:::caution

Use a dedicated bucket for your Harness CI cache operations. Do not save files to the bucket manually. The Restore Cache operation will fail if the bucket includes any files that do not have a Harness cache key.

:::

## Add Save Cache to S3 step

Here is a YAML example of a **Save Cache to S3** step.

```yaml
              - step:
                  type: SaveCacheS3
                  name: Save Cache to S3
                  identifier: Save_Cache_to_S3
                  spec:
                    connectorRef: AWS_Connector
                    region: us-east-1
                    bucket: your-s3-bucket
                    key: cache-{{ checksum filePath1 }} # example cache key based on file checksum
                    sourcePaths:
                      - directory1 # example first directory to cache
                      - directory2 # example second directory to cache
                    archiveFormat: Tar
...
```

For details about this step's settings, go to [Save Cache to S3 step settings](../../ci-technical-reference/save-cache-to-s-3-step-settings.md).

> Pipeline steps within a stage share the same [workspace](../../ci-technical-reference/ci-stage-settings.md#workspace).
> 
> You can optionally [share paths](../../ci-technical-reference/ci-stage-settings.md#share-paths) outside the workspace between steps in your stage by setting `spec.sharedPaths`.

Reviewer: ```suggestion
Pipeline steps within a stage share the same [workspace](../../ci-technical-reference/ci-stage-settings.md#workspace). You can optionally [share paths](../../ci-technical-reference/ci-stage-settings.md#share-paths) outside the workspace between steps in your stage by setting `spec.sharedPaths`.
```

```

Add **Save Cache to S3** step to your stage after steps that build and test your code, as shown in this diagram:

```mermaid
graph TD
  accTitle: Save Cache to S3
  accDescr: Diagram showing Save Cache to S3 step usage in a single stage

  A(Restore Cache From S3)-- Build and Test Your Code --> B
  B(<strong>Save Cache to S3</strong>)
```

## Add Restore Cache From S3 step

Here is a YAML example of a **Restore Cache From S3** step.

```yaml
              - step:
                  type: RestoreCacheS3
                  name: Restore Cache From S3
                  identifier: Restore_Cache_From_S3
                  spec:
                    connectorRef: AWS_Connector
                    region: us-east-1
                    bucket: your-s3-bucket
                    key: cache-{{ checksum filePath1 }} # example cache key based on file checksum
                    archiveFormat: Tar
...
```

The `spec.key` value in this step must match the `spec.key` value in your **Save Cache to S3** step.

For details about this step's settings, go to [Restore Cache from S3 step settings](../../ci-technical-reference/restore-cache-from-s-3-step-settings.md).

> Add **Restore Cache From S3** step to your stage before steps that build and test your code, as shown in this diagram:

Reviewer: ```suggestion
Add **Restore Cache From S3** step to your stage before steps that build and test your code, as shown in the following diagram:
```

```mermaid
graph TD
  accTitle: Restore Cache From S3
  accDescr: Diagram showing Restore Cache From S3 step usage in a single stage
  A(<strong>Restore Cache From S3</strong>)-- Build and Test Your Code --> B
  B(Save Cache to S3)
```

## Languages

Your cache key and paths will differ depending on your language.

```mdx-code-block
<Tabs>
<TabItem value="Go">
```

> [Go](https://go.dev/) pipelines should reference `go.sum` for `spec.key` in **Save Cache to S3** and **Restore Cache From S3** steps.

Reviewer: ```suggestion
[Go](https://go.dev/) pipelines must reference `go.sum` for `spec.key` in **Save Cache to S3** and **Restore Cache From S3** steps, for example:
```

```yaml
                  spec:
                    key: cache-{{ checksum "go.sum" }}
```

> `spec.sourcePaths` should include `/go/pkg/mod` and `/root/.cache/go-build` in the **Save Cache to S3** step.

Reviewer: ```suggestion
`spec.sourcePaths` must include `/go/pkg/mod` and `/root/.cache/go-build` in the **Save Cache to S3** step, for example:
```

```yaml
                  spec:
                    sourcePaths:
                    - /go/pkg/mod
                    - /root/.cache/go-build
```

```mdx-code-block
</TabItem>

<TabItem value="Node.js">
```

> [Npm](https://www.npmjs.com/) pipelines should reference `package-lock.json` for `spec.key` in **Save Cache to S3** and **Restore Cache From S3** steps.

Reviewer: ```suggestion
[Npm](https://www.npmjs.com/) pipelines must reference `package-lock.json` for `spec.key` in **Save Cache to S3** and **Restore Cache From S3** steps, for example:
```

```yaml
                  spec:
                    key: cache-{{ checksum "package-lock.json" }}
```

> [Yarn](https://yarnpkg.com/) pipelines should reference should reference `yarn.lock` for `spec.key` in **Save Cache to S3** and **Restore Cache From S3** steps.

Reviewer: ```suggestion
[Yarn](https://yarnpkg.com/) pipelines must reference `yarn.lock` for `spec.key` in **Save Cache to S3** and **Restore Cache From S3** steps, for example:
```

```yaml
                  spec:
                    key: cache-{{ checksum "yarn.lock" }}
```

> `spec.sourcePaths` should include `node_modues` in the **Save Cache to S3** step.

Reviewer: ```suggestion
`spec.sourcePaths` must include `node_modues` in the **Save Cache to S3** step, for example:
```

```yaml
                  spec:
                    sourcePaths:
                    - node_modules
```

```mdx-code-block
</TabItem>

<TabItem value="Maven">
```

> [Maven](https://maven.apache.org/) pipelines should reference `pom.xml` for `spec.key` in **Save Cache to S3** and **Restore Cache From S3** steps.

Reviewer: ```suggestion
[Maven](https://maven.apache.org/) pipelines must reference `pom.xml` for `spec.key` in **Save Cache to S3** and **Restore Cache From S3** steps, for example:
```

```yaml
                  spec:
                    key: cache-{{ checksum "pom.xml" }}
```

> `spec.sourcePaths` should include `/root/.m2` in the **Save Cache to S3** step.

Reviewer: ```suggestion
`spec.sourcePaths` must include `/root/.m2` in the **Save Cache to S3** step, for example:
```

```yaml
                  spec:
                    sourcePaths:
                    - /root/.m2
```

```mdx-code-block
</TabItem>
</Tabs>
```

## Multi-Stage Pipelines

> Each stage runs in an isolated environment, caching can be used to pass data from one stage to the next.

Reviewer: ```suggestion
Because each stage runs in an isolated environment, you can use caching to pass data from one stage to the next.
```

> This diagram illustrates cache usage across two stages.

Reviewer: ```suggestion
The following diagram illustrates cache usage across two stages.
```
